File: src/network_adapter.cpp
/* Copyright 2017-2018 All Rights Reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *  Eunsoo Park (esevan.park@gmail.com)
 *  Injung Hwang (sinban04@gmail.com)
 *  
 * [Contact]
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0(the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <network_adapter.h>

#include <dbug_log.h>
#include <network_manager.h>
#include <segment_manager.h>

#include <string.h>
#include <errno.h>
#include <netinet/in.h>

#include <functional>

namespace cm {

NetworkAdapter::~NetworkAdapter() {
  if (at == kATUninitialized)
    return;

  dev_off();

  if (at & kATCtrl)
    NetworkManager::get_instance()->remove_control_adapter(this);
  else
    NetworkManager::get_instance()->remove_data_adapter(this);
}

/* 
 * Try to make an adapter in the target state
 */
void NetworkAdapter::dev_switch(DevState stat, DevStatCb cb) {
  switch (stat) {
    case kDevOn:
      /* Not implemented
      device_on_cb = cb;
      std::thread(std::bind(&NetworkAdapter::dev_on, this)).detach();
      break;
      */
    case kDevOff:
      /* Not implemented
      device_off_cb = cb;
      std::thread(std::bind(&NetworkAdapter::dev_off, this)).detach();
      */
      LOG_ERR("Not implemented bug");
      break;
    case kDevDiscon:
      if (this->stat == kDevDiscon) {
        LOG_WARN("Already disconnected");
        return;
      }

      if (this->stat == kDevDisconnecting) {
        LOG_WARN("Device is busy");
        return;
      }
      this->stat = kDevDisconnecting;
      close_connection_cb = cb;
      std::thread(std::bind(&NetworkAdapter::_close, this)).detach();
      break;
    case kDevCon:
      // Can connect the device only if it's in the disconnect state
      if (this->stat > kDevDiscon) {
        LOG_WARN("Cannot connect this adapter: stat(%d)", this->stat);
        return;
      }

      this->stat = kDevConnecting;
      make_connection_cb = cb;
      std::thread(std::bind(&NetworkAdapter::_connect, this)).detach();
      break;
  }
}

void NetworkAdapter::set_data_adapter(void) {
  // Check if the adapter is initialized
  if (at == kATInitialized) {
    LOG_WARN("Already initialized: %d", at & kATCtrl);
    return;
  }
  // Install the adapter as the data adapter
  NetworkManager::get_instance()->install_data_adapter(this);
  at |= kATInitialized;
}

/*  This function sets the adapter as a control adapter
 */
void NetworkAdapter::set_control_adapter(void) {
  // Check if the adapter is initialized
  if (at & kATInitialized) {
    LOG_WARN("Already initialized: %d", at & kATCtrl);
    return;
  }
  
  // Check if the adapter is controllable
  if ((at & kATCtrlable) == 0) {
    LOG_WARN("Not controllable adapter %d", dev_id);
    return;
  }
  // Install the control adapter
  at |= kATCtrl | kATInitialized;
  NetworkManager::get_instance()->install_control_adapter(this);

}

void NetworkAdapter::join_threads(){
  SegmentManager *sm = SegmentManager::get_instance();
  sm->set_is_wfd_on(0);

  LOG_VERB("wait for the sender thread to end\n");

  sender_semaphore = 1;
//  sender_start.notify_one();
//  if(th_sender->joinable()) {
//    th_sender->join(); 
//  }
  sender_semaphore = 0;

  LOG_VERB("wait for the recver thread to end\n");
  recver_semaphore = 1;
//  recver_start.notify_one();
//  if(th_recver->joinable()) {
//    th_recver->join(); 
//  }
  recver_semaphore = 0;
  
 
  LOG_VERB("Joined the sender & recver thread and delete them\n");
//  delete th_sender;
//  delete th_recver;
//
//  th_sender = NULL;
//  th_recver = NULL;

  /* Send control message to turn off the working data adapter */
  NetworkManager *nm = NetworkManager::get_instance();
  unsigned char buf[512];
  buf[0] = kCtrlReqDecr;
  uint16_t ndev_id = htons(nm->decreasing_adapter_id);
  memcpy(buf+1, &ndev_id, 2);
  nm->send_control_data((const void *)buf, 3);
}

bool NetworkAdapter::delete_threads(){
  std::thread(std::bind(&NetworkAdapter::join_threads, this)).detach();

  return true;
}

DevState NetworkAdapter::get_stat() {
  return stat;
}

void NetworkAdapter::dev_on(void) {
  std::unique_lock<std::mutex> lck(dev_on_lock);
  if (stat == kDevOff) {
    bool res = this->device_on();
    if (res)
      stat = kDevOn;
    else
      stat = kDevOff;
  }

  if (device_on_cb) device_on_cb(stat);

  device_on_cb = NULL;
}

void NetworkAdapter::dev_off(void) {
  std::unique_lock<std::mutex> lck(dev_off_lock);
  if (stat == kDevOn) {
    bool res = this->device_off();
    if (res)
      stat = kDevOff;
    else
      stat = kDevOn;
  }

  if (device_off_cb) device_off_cb(stat);
  device_off_cb = NULL;
}
/*  This function is connecting function of the adapter as the name says.
 *  This is run by the independent thread created in the dev_switch() function. 
 */
void NetworkAdapter::_connect(void) {

  printf("connect thread created! tid: %d\n", (unsigned int)syscall(224));
  // It should be in the connecting state, when it really tries to connect
  if (stat != kDevConnecting){
    return;
  }
  /*  make_connection() function is the function defined in each network adapter
   */
  bool res = make_connection();
  
  if (res) 
  {
    /*  Only if it's data adapter, run the sender & recver thread.
     *  Control adapter uses other thread (control_recver_thread)
     */
    if ((at & kATCtrl) == 0) {  // Data Adapter
      LOG_VERB("Data adapter connected");
      th_sender =
          new std::thread(std::bind(&NetworkAdapter::run_sender, this));
      th_recver =
          new std::thread(std::bind(&NetworkAdapter::run_recver, this));
      //th_sender->detach();
      //th_recver->detach();
    } else { // Control Adapter
      LOG_VERB("Control adapter connected");
      th_sender = NULL;
      th_recver = NULL;
    }

    stat = kDevCon;
  }
  else {
    stat = kDevDiscon;
    LOG_VERB("NetworkAdapter::_connect: Not connected!\n");
  }
  if (make_connection_cb){
    LOG_VERB("connect callback called!\n");
    make_connection_cb(stat);
  }
  make_connection_cb = NULL;
}

void NetworkAdapter::_close(void) {
  __FUNCTION_ENTER__;
  if (stat != kDevDisconnecting)
    return;

  bool res = close_connection();
  if (res) {
    LOG_VERB("connection closed\n");
    stat = kDevDiscon;
  } else {
    stat = kDevCon;
  }

  join_threads();

//  if((at & kATCtrl) == 0) { //Data adapter
//    while(1){
//    if(th_sender == NULL && th_recver == NULL){
//      LOG_VERB("No thread exists\n");
//    
//      break;
//    
//    }
//    else {
//      LOG_WARN("Still threads exist. Let's wait more\n"); 
//
//    }
//
//    }//end_while
//
//  } else { // Control Adapter
//    LOG_ERR("Try to turn off the control adapter\n");
//    exit(1);
//  }

  if (close_connection_cb) {
    LOG_VERB("Call close_connection callback\n");
    close_connection_cb(stat);
  } else{
    LOG_VERB("No callback\n");
  }
  close_connection_cb = NULL;
}

void NetworkAdapter::return_sending_failed_packet(void *seg) {
  Segment *to_send = reinterpret_cast<Segment *>(seg);

  SegmentManager::get_instance()->failed_sending(to_send);
}

void NetworkAdapter::run_sender(void) {
  printf("sender thread created! tid: %d\n", (unsigned int)syscall(224));
  SegmentManager *sm = SegmentManager::get_instance();

  if(net_dev_type == kWifiDirect) {
    sm->set_is_wfd_on(1);
    LOG_VERB("WiFi Direct sender thread start working\n");
  } else if (net_dev_type == kBluetooth) {
    LOG_VERB("Bluetooth sender thread start working\n"); 
  }

 while (true) { 
    if (net_dev_type == kBluetooth) {
      while (sm->get_is_wfd_on() == 1) {
        LOG_DEBUG("Wifi-direct is on. stop BT sender thread\n");
        /*
        std::unique_lock<std::mutex> lck1(sender_lock);
        sender_start.wait(lck1); 
        */
        sleep(1);
        LOG_DEBUG("wfd on. is_wfd_on: %d\n", sm->get_is_wfd_on());
      }
    }

    if (sender_semaphore == 1) {
      LOG_VERB("sender semaphore is 1. stop sender thread\n");
      if(net_dev_type == kWifiDirect) {
        LOG_DEBUG("wfd off. is_wfd_on: %d\n", sm->get_is_wfd_on());
      }
      break;
    }

    Segment *to_send;

    /* At first, dequeue a segment from failed sending queue */
    to_send = sm->get_failed_sending();

    /* If there is no failed segment, dequeue from send queue */
    if (likely(to_send == NULL)){
      to_send = sm->dequeue(kSegSend);
    }

    int len = kSegHeaderSize + kSegSize;
    const void *data = to_send->data;
    //LOG_VERB("to_send seq_no: %d", to_send->seq_no);
    bool res = this->send(data, len); 
    if (res < 0) {
      LOG_WARN("Sending failed at %s (%d; %s)", dev_name, errno, strerror(errno));
      return_sending_failed_packet(to_send);
      break;
    }
    sm->free_segment(to_send);
  }

 LOG_VERB("Sender thread ends\n");

  // dev_switch(kDevDiscon, NULL);
}

void NetworkAdapter::run_recver(void) {
  printf("recever thread created! tid: %d\n", (unsigned int)syscall(224));
  SegmentManager *sm = SegmentManager::get_instance();
  Segment *free_seg = sm->get_free_segment();

  if(net_dev_type == kWifiDirect){
    LOG_VERB("WiFi Direct recver thread start working\n");
  } else if (net_dev_type = kBluetooth){
    LOG_VERB("Bluetooth recver thread start working\n"); 
  }


  while (true) {
    if(net_dev_type == kBluetooth){
      while(sm->get_is_wfd_on() == 1){
        LOG_DEBUG("Wifi-direct is on. stop BT recver thread\n");
        /*
        std::unique_lock<std::mutex> lck2(recver_lock);
        recver_start.wait(lck2);
        */
        sleep(1);
      }
    }
    
   if(recver_semaphore == 1){
     LOG_WARN("recver semaphore is 1. stop recver thread\n");
     break;
   }

    void *buf = reinterpret_cast<void *>(free_seg->data);
    int len = kSegSize + kSegHeaderSize;
    LOG_DEBUG("Receiving...");
    int res = this->recv(buf, len);
    if (res < len) {
      LOG_WARN("Recving failed at %s (%s)", dev_name, strerror(errno));
      sm->free_segment(free_seg);
      break;
    }

    uint32_t net_seq_no, net_flag_len;
    memcpy(&net_seq_no, buf, sizeof(uint32_t));
    memcpy(&net_flag_len,
           (reinterpret_cast<uint8_t *>(buf)+4), sizeof(uint32_t));
    free_seg->seq_no = ntohl(net_seq_no);
    free_seg->flag_len = ntohl(net_flag_len);

    LOG_DEBUG("Recved:%d/%d(%d)", free_seg->seq_no, free_seg->flag_len, res);
    sm->enqueue(kSegRecv, free_seg);
    free_seg = sm->get_free_segment();
  }
 
  LOG_VERB("Recver thread ends\n");

  //dev_switch(kDevDiscon, NULL);
}

void NetworkAdapter::set_controllable(void) {
  at = at | kATCtrlable;
}

void NetworkAdapter::send_ctrl_msg(const void *buf, int len) {
  uint8_t req = kCtrlReqPriv;
  uint16_t net_dev_id = htons(this->dev_id);
  uint32_t net_len = htonl(len);

  if ((at & kATCtrl) == kATCtrl) {
    LOG_WARN("Cannot transfer private data in control adapter");
    return;
  }

  NetworkManager *nm = NetworkManager::get_instance();
  nm->send_control_data(&req, 1);
  nm->send_control_data(&net_dev_id, 2);
  nm->send_control_data(&net_len, 4);
  nm->send_control_data(buf, len);
}

int NetworkAdapter::send(const void *buf, size_t len) {
  int ret;
  ret = this->send_impl(buf, len);
  if(ret > 0) {
    this->mSendDataSize.add((int)len);
  }
  return ret;
}

int NetworkAdapter::recv(void *buf, size_t len) {
  int ret;
  ret = this->recv_impl(buf, len);
  if(ret > 0) {
    this->mReceiveDataSize.add((int)len);
  }
  return ret;
}

}  /* namespace cm */

File: android/app/src/main/java/com/redcarrottt/sc/NetworkAdapter.java
package com.redcarrottt.sc;

/* Copyright (c) 2017-2018. All rights reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *  Eunsoo Park (esevan.park@gmail.com)
 *  Injung Hwang (sinban04@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.os.Handler;

import com.redcarrottt.testapp.Logger;

import java.nio.ByteBuffer;

import static com.redcarrottt.sc.SegmentManager.kSegHeaderSize;
import static com.redcarrottt.sc.SegmentManager.kSegRecv;
import static com.redcarrottt.sc.SegmentManager.kSegSend;
import static com.redcarrottt.sc.SegmentManager.kSegSize;

public abstract class NetworkAdapter {
    static public final int kDevOff = 0;
    static public final int kDevOn = 1;
    static public final int kDevDiscon = 2;
    static public final int kDevDisconnecting = 3;
    static public final int kDevConnecting = 4;
    static public final int kDevCon = 5;

    static public final int kATuninitialized = 0;
    static public final int kATInitialized = 1;
    static public final int kATCtrlable = 2;
    static public final int kATCtrl = 4;

    static public final int kBluetooth = 1;
    static public final int kWifi = 2;
    static public final int kWifiDirect = 3;

    private int sender_semaphore = 0;
    private int recver_semaphore = 0;
    private Object sender_lock;
    private Object recver_lock;

    public abstract boolean device_on();

    public abstract boolean device_off();

    public abstract boolean make_connection();

    public abstract boolean close_connection();

    public abstract int send(byte[] buf, int len);

    public abstract int recv(byte[] buf, int len);

    public abstract void on_control_recv(byte[] buf, int len);

    private class SenderThread extends Thread {
        public void run() {
            SegmentManager sm = SegmentManager.getInstance();
            if (dev_type == kWifiDirect) {
                sm.wfd_state = 1;
            }

            if (dev_type == kBluetooth) {
                Logger.print(tag, "start bluetooth sender thread");
            } else if (dev_type == kWifiDirect) {
                Logger.print(tag, "start WiFi direct sender thread");
            }

            while (true) {
                if (dev_type == kBluetooth) {
                    while (sm.wfd_state == 1) {
                        try {
                            //Logger.print(tag, "WFD is on. stop bluetooth
                            // sender thread");
                            sleep(1000);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                }

                if (sender_semaphore == 1) {
                    Logger.print(tag, "sender semaphore is 1. stop sender " +
                            "thread");
                    if (dev_type == kWifiDirect) {
                        sm.wfd_state = 0;
                    }
                    break;
                }

                Segment to_send = sm.dequeue(kSegSend);
                if (to_send != null) {
                    int len = kSegSize + kSegHeaderSize;
                    byte[] data = to_send.data;
                    int res = send(data, len);
                    if (res < len) {
                        sm.enqueue(kSegSend, to_send);
                        if (res <= 0) break;
                    }
                    sm.free_segment(to_send);
                }
            }
            synchronized (sender_lock) {
                sender_lock.notify();
            }
        }
    }

    private class RecverThread extends Thread {
        public void run() {
            SegmentManager sm = SegmentManager.getInstance();
            Segment free_seg = sm.get_free_segment();

            if (dev_type == kBluetooth) {
                Logger.print(tag, "start bluetooth recver thread");
            } else if (dev_type == kWifiDirect) {
                Logger.print(tag, "start WiFi direct recver thread");
            }

            while (true) {
                /*
                if(dev_type == kBluetooth){

                    while(sm.wfd_state == 1){
                        try {
                            Logger.print(tag, "WFD is on. stop bluetooth
                            recver thread");
                            sleep(1000);
                        } catch (Exception e){
                            e.printStackTrace();
                        }
                    }
                }*/

                if (recver_semaphore == 1) {
                    Logger.print(tag, "recver semaphore is 1. stop the " +
                            "recver thread");
                    break;
                }

                int len = kSegSize + kSegHeaderSize;
                //Logger.print(tag, "Recving thread start:"+Integer.toString
                // (dev_id));
                int res = recv(free_seg.data, len);
                if (res < len) {
                    Logger.print(tag, "Recving failed");
                    sm.free_segment(free_seg);

                    if (res <= 0) {
                        break;
                    }
                } else {
                    //Logger.print(tag, "Recved : " + Short.toString(dev_id));
                }

                ByteBuffer buffer = ByteBuffer.allocate(4);
                buffer.put(free_seg.data, 0, 4);
                free_seg.seq_no = buffer.getInt(0);
                //Logger.print(tag, "Recved Seq No : " + Integer.toString
                // (free_seg.seq_no));

                buffer = ByteBuffer.allocate(4);
                buffer.put(free_seg.data, 4, 4);
                free_seg.flag_len = buffer.getInt(0);
                //Logger.print(tag, "Recved flag_len: " + Integer.toString
                // (free_seg.flag_len));

                sm.enqueue(kSegRecv, free_seg);
                free_seg = sm.get_free_segment();
            }
            try {
                sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            synchronized (recver_lock) {
                recver_lock.notify();
            }
        }
    }

    private class ConnectingThread extends Thread {
        Handler mHandler;

        ConnectingThread(Handler handler) {
            mHandler = handler;
        }

        public void run() {
            if (stat != kDevConnecting) return;

            boolean res = make_connection();

            if (res) {
                if ((at & kATCtrl) == 0) {
                    if (send_thread != null) send_thread.interrupt();

                    send_thread = new SenderThread();
                    send_thread.start();

                    if (recv_thread != null) recv_thread.interrupt();

                    recv_thread = new RecverThread();
                    recv_thread.start();
                } else {
                    if (send_thread != null) send_thread.interrupt();

                    if (recv_thread != null) recv_thread.interrupt();

                    send_thread = null;
                    recv_thread = null;
                }

                stat = kDevCon;
            } else {
                stat = kDevDiscon;
            }

            mHandler.sendEmptyMessage(stat);
        }
    }

    private class ClosingThread extends Thread {
        Handler mHandler;

        ClosingThread(Handler handler) {
            mHandler = handler;
        }

        public void run() {
            if (stat != kDevDisconnecting) return;

            if (send_thread != null) {
                sender_semaphore = 1;
                try {
                    synchronized (sender_lock) {
                        sender_lock.wait();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                send_thread.interrupt();
                send_thread = null;
                Logger.print(tag, "sender thread stopped ");
                sender_semaphore = 0;
            }
            if (recv_thread != null) {
                recver_semaphore = 1;
                try {
                    synchronized (recver_lock) {
                        recver_lock.wait();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }

                recv_thread.interrupt();
                recv_thread = null;
                Logger.print(tag, "recver thread stopped ");
                recver_semaphore = 0;
            }

            boolean res = close_connection();

            if (res) stat = kDevDiscon;
            else stat = kDevCon;

            if (mHandler != null) mHandler.sendEmptyMessage(stat);
        }
    }

    private String tag = "NetworkAdapter";

    private SenderThread send_thread;
    private RecverThread recv_thread;
    private ConnectingThread connect_thread;
    private ClosingThread close_thread;

    public int dev_type;
    public short dev_id = 0;
    public int stat;
    private int at;

    NetworkAdapter() {
        at = kATuninitialized;
        stat = kDevDiscon;
        dev_id = 0;

        send_thread = null;
        recv_thread = null;
        connect_thread = null;
        close_thread = null;

        sender_lock = new Object();
        recver_lock = new Object();
        dev_type = 0;
    }

    public final void set_controllable() {
        at |= kATCtrlable;
    }

    public final void set_data_adapter() {
        if ((at & kATInitialized) == kATInitialized) {
            Logger.print(tag, "Already initialized");
            return;
        }

        Core.getInstance().install_data_adapter(this);
        at |= kATInitialized;
    }

    public final void set_control_adapter() {
        if ((at & kATInitialized) == kATInitialized) {
            Logger.print(tag, "Already initialized");
            return;
        }

        if ((at & kATCtrlable) == 0) {
            Logger.print(tag, "Not controllable adapter");
            return;
        }
        at |= kATCtrl | kATInitialized;

        Core.getInstance().install_control_adapter(this);
    }

    public void connect(Handler handler) {
        if (at == kATuninitialized) {
            Logger.print(tag, "The adapter has not been installed to " +
                    "core");
            return;
        }

        if (stat != kDevDiscon) {
            Logger.print(tag, "The adapter is not disconnected");
            return;
        }

        stat = kDevConnecting;
        if (connect_thread != null) {
            connect_thread.interrupt();
        }

        connect_thread = new ConnectingThread(handler);
        connect_thread.start();
    }

    public void close(Handler handler) {
        if (stat != kDevCon) {
            Logger.print(tag, "The adapter is not connected");
            return;
        }

        Logger.print(tag, "close adapter");
        stat = kDevDisconnecting;
        if (close_thread != null) {
            close_thread.interrupt();
        }

        close_thread = new ClosingThread(handler);
        close_thread.start();
    }

    public int get_stat() {
        return stat;
    }

    public void send_ctrl_msg(byte[] data, int len) {
        if ((at & kATCtrl) == kATCtrl) {
            Logger.print(tag, "Cannot transfer private data in control " +
                    "adapter");
            return;
        }

        Core core = Core.getInstance();
        ByteBuffer buffer = ByteBuffer.allocate(1);
        final int kCtrlReqPriv = 2;
        buffer.put((byte) kCtrlReqPriv);

        core.sendControlMessage(buffer.array(), 1);

        buffer = ByteBuffer.allocate(2);
        buffer.putShort(dev_id);
        core.sendControlMessage(buffer.array(), 2);

        buffer = ByteBuffer.allocate(4);
        buffer.putInt(len);
        core.sendControlMessage(buffer.array(), 4);

        core.sendControlMessage(data, len);
    }
}
